fem/lininteg_domain.hpp

```diff
@@ -0,0 +1,263 @@
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+// Internal header, included only by .cpp files
+
+#include "fem.hpp"
+
+#include "../fem/kernels.hpp"
+#include "../general/forall.hpp"
+#include "../linalg/kernels.hpp"
+
+namespace mfem
+{
+
+namespace internal
+{
+
+namespace linearform_extension
+{
+
+/// Signature of the kernels used for linear form extension
+using LinearFormExtensionKernel_f = void (*)(const int vdim,
+                                             const bool byVDIM,
+                                             const int ND,
+                                             const int NE,
+                                             const int d,
+                                             const int q,
+                                             const int *markers,
+                                             const double *b,
+                                             const double *g,
+                                             const int *idx,
+                                             const double *J,
+                                             const double *detJ,
+                                             const double *weights,
+                                             const Vector &coeff,
+                                             double *output);
+
+/// Signature of the function used to compute the quadrature order
+using GetOrder_f = std::function<int(int)>;
+
+/// Internal helper function to get the integration rule
+inline const IntegrationRule *GetIntRuleFromOrder(const FiniteElementSpace &fes,
+                                                  const IntegrationRule *IntRule,
+                                                  const GetOrder_f &qorder_fct)
+{
+   const FiniteElement &fe = *fes.GetFE(0);
+   const int qorder = qorder_fct(fe.GetOrder());
+   const Geometry::Type geom_type = fe.GetGeomType();
+   return IntRule ? IntRule : &IntRules.Get(geom_type, qorder);
+}
+
+/// Internal helper function to encode the ID of a LinearFormExtension kernel
+/// The ID is equal to: (dim << 8) |
+///                     (1D number of degrees of freedom << 4) |
+///                     (1D number of quadrature points)
+inline int GetKernelId(const FiniteElementSpace &fes,
+                       const IntegrationRule *ir)
+{
+   Mesh *mesh = fes.GetMesh();
+   const int dim = mesh->Dimension();
+   const FiniteElement &el = *fes.GetFE(0);
+   const DofToQuad &maps = el.GetDofToQuad(*ir, DofToQuad::TENSOR);
+   const int d = maps.ndof;
+   const int q = maps.nqpt;
```

Comment:
Many kernels could be retrofitted I agree, but depending on the timing of the new interfaces, too early adoption and altering everything now appears to me as a bad idea.

Reply:
The `16` doesn't have to be replaced by a global definition, just a self documented name, but it's not that important. What is important is the `MFEM_ASSERT` since currently it's undocumented and unchecked that `q` and `d` have to be smaller than `16`.

Reply:
Added `MFEM_VERIFY(q < 32, "Unsupported quadrature order!");`

Reply:
I may be missing something but you are encoding `q` on 4 bits, and `d` on 4 bits, so each of them has to be smaller than `16`, no?
```c++
MFEM_VERIFY(q < 16, "Unsupported quadrature order!");
MFEM_VERIFY(d < 16, "Unsupported polynomial order!");
```

Reply:
`done`

Reply:
Where?